File: Src/Runtime/Strings/Split.cpp
/*
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #                           Typed                     # *
 * #                   A text formatting DSL             # *
 * #                                                     # *
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #         Created by Rodrigo R. & Contributors        # *
 * #         Released under the Apache License 2.0       # *
 * #            Check LICENSE.MD for more info           # *
 * #                                                     # *
 * #-----------------------------------------------------# *
*/

//
// Created by Rodrigo on 6/15/26.
//

#include "ADT/List/Object.h"
#include "Split.h"

#include "Support/Runtime/AccessString.h"

using namespace Typed;
using namespace Typed::Runtime;
using namespace Typed::Runtime::Strings;

ADT::Runtime::Object Strings::Split(
    ADT::List::Object &args,
    ADT::Lang::AST *_
)
{
    auto &obj = args[0];
    auto str = Support::Runtime::AccessString(obj);
    auto delimiter = Support::Runtime::AccessString(args[1]);

    auto result = ADT::List::DynamicObject::Make();
    std::string current;
    size_t i = 0;

    while (i < str.size())
    {
        bool match = true;

        if (i + delimiter.size() <= str.size())
        {
            for (auto j = 0; j < delimiter.size(); ++j)
            {
                if (str[i + j] != delimiter[j])
                {
                    match = false;
                    break;
                }
            }
        }
        else
        {
            match = false;
        }

        if (match)
        {
            result->emplace_back(
                ADT::Runtime::ObjectType::String,
                std::move(current)
            );

            i += delimiter.size();
        }
        else
        {
            current.push_back(str[i]);
            ++i;
        }
    }

    result->emplace_back(
        ADT::Runtime::ObjectType::String,
        std::move(current)
    );

    return {
        ADT::Runtime::ObjectType::List,
        std::move(result)
    };
}

File: src/runtime/strings/split.rs
/*
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #                           Typed                     # *
 * #                   A text formatting DSL             # *
 * #                                                     # *
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #         Created by Rodrigo R. & Contributors        # *
 * #         Released under the Apache License 2.0       # *
 * #            Check LICENSE.MD for more info           # *
 * #                                                     # *
 * #-----------------------------------------------------# *
*/
use std::cell::RefCell;
use crate::adt::lang::{ASTValue, RuntimeArguments, AST};
use crate::adt::result::ExecutionTupleResult;
use crate::adt::runtime::{HashableObject, NonHashableObject, Object, RuntimeList};
use crate::support::runtime::execution::continue_execution;
use crate::support::runtime::object::get_string;

pub fn split<'a>(
    args: RuntimeArguments<'a>,
    trace: &AST<'a>
)
    -> ExecutionTupleResult<'a>
{
    let origin = args.get(0).unwrap();
    let delim = args.get(1).unwrap();

    let origin = get_string(origin, trace)?;
    let delim = get_string(delim, trace)?;

    let result: RuntimeList = RuntimeList::new(RefCell::new(vec![]));
    let parts: Vec<&str> = origin.split(delim).collect();

    for part in parts {
        let mut result = result.borrow_mut();
        let obj = Object::Hashable(HashableObject::String(
            ASTValue::Owned(
                part.to_string()
            )
        ));

        result.push(obj);
    }

    continue_execution(
        Object::NonHashable(
            NonHashableObject::List(
                result
            )
        )
    )
}